interfaces/builtin/gpio_chardev.go

```diff
@@ -0,0 +1,230 @@
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/features"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/systemd"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/snap"
+)
+
+// TODO: Snapd should validate the correctness of slot declarations
+// (i.e. lines across slots are unique) when installing the gadget
+// snap e.g. when validating snap.yaml
+
+// https://docs.kernel.org/userspace-api/gpio/chardev.html
+// https://docs.kernel.org/admin-guide/gpio/gpio-aggregator.html
+const gpioChardevSummary = `allows access to specific GPIO chardev lines`
+
+const gpioChardevBaseDeclarationSlots = `
+  gpio-chardev:
+    allow-installation:
+      slot-snap-type:
+        - gadget
+    deny-auto-connection: true
+`
+
+var gpioChardevConnectedSlotKmod = []string{
+	"gpio-aggregator",
+}
+
+// XXX: What should be the limit on max range.
+const maxLinesCount = 65536
+
+func parseLineToken(lineToken string) ([]uint64, error) {
+	if !strings.Contains(lineToken, "-") {
+		line, err := strconv.ParseUint(lineToken, 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		return []uint64{line}, nil
+	}
+	// Parse line range e.g. 2-5
+	tokens := strings.SplitN(lineToken, "-", 2)
+	if len(tokens) != 2 {
+		return nil, fmt.Errorf("invalid line range %q", lineToken)
+	}
+	first, err := strconv.ParseUint(tokens[0], 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("invalid line range %q: %w", lineToken, err)
+	}
+	last, err := strconv.ParseUint(tokens[1], 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("invalid line range %q: %w", lineToken, err)
+	}
+	if last <= first {
+		return nil, fmt.Errorf("invalid line range %q: range end has to be larger than range start", lineToken)
+	}
+	if last-first+1 > maxLinesCount {
+		return nil, fmt.Errorf("invalid line range %q: range cannot be more than %d lines", lineToken, maxLinesCount)
+	}
+	lines := make([]uint64, 0, last-first+1)
+	for i := first; i <= last; i++ {
+		lines = append(lines, i)
+	}
+	return lines, nil
+}
+
+func validateLines(linesAttr string) error {
+	tokens := strings.Split(linesAttr, ",")
+
+	lines := make(map[uint64]bool, len(tokens))
+	for _, token := range tokens {
+		tokenLines, err := parseLineToken(token)
+		if err != nil {
+			return err
+		}
+		for _, line := range tokenLines {
+			if _, exists := lines[line]; exists {
+				return fmt.Errorf(`duplicate line found "%d"`, line)
+			}
+			lines[line] = true
+		}
+	}
+	if len(lines) > maxLinesCount {
+		return fmt.Errorf("number of lines cannot be more than %d", maxLinesCount)
+	}
+
+	return nil
+}
+
+type gpioChardevInterface struct {
+	commonInterface
+}
+
+// BeforePrepareSlot checks validity of the defined slot.
+func (iface *gpioChardevInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	var sourceChip []string
+	// "source-chip" attribute is mandatory.
+	if err := slot.Attr("source-chip", &sourceChip); err != nil {
```

Comment:
wort checking that the entries aren't empty strings, or duplicates

Reply:
good point, thanks!